File: src/db-gwent-server-kotlin/src/main/kotlin/com/entities/PointGroupModifier.kt
package com.entities

class PointGroupModifier(
    val value: Int = 0,
    val type: PointGroupModifierType,
    val id: String = "",
    val getOriginalValue: () -> Int) {

    fun modify(points: Int): Int {
        return when (type) {
            PointGroupModifierType.Add -> points + value
            PointGroupModifierType.Multiply -> points * value
            PointGroupModifierType.MultiplyOriginal -> points + getOriginalValue()
            PointGroupModifierType.Constant -> value
            else -> points
        }
    }
}

enum class PointGroupModifierType {
    Add, Multiply, MultiplyOriginal, Constant
}

File: src/db-gwent-server-scala/src/main/scala/com/entities/PointGroupModifier.scala
package com.entities

case class PointGroupModifier(
  value: Int = 0,
  $type: PointGroupModifierType,
  id: String = ""):

  def modify(points: Int): Int =
    $type match
      case PointGroupModifierType.Add => points + value
      case PointGroupModifierType.Multiply => points * value
      case PointGroupModifierType.Constant => value

enum PointGroupModifierType:
  case Add, Multiply, Constant
